### PR TITLE
readers.stac implementation

### PR DESCRIPTION
**Pull request by hobu**
A number of tweaks and updates of #3895 due to it not being editable.

### CODE REVIEW

File: io/StacReader.cpp [hunk 1]
```diff
@@ -0,0 +1,521 @@
...
+        {
+            log()->get(LogLevel::Debug) << "Processing extension " << extSchemaUrl << std::endl;
+            std::string schemaStr = m_arbiter->get(extSchemaUrl);
+            NL::json schemaJson = NL::json::parse(schemaStr);
+            val.set_root_schema(schemaJson);
+            val.validate(stacJson);
+        }
+    }
+    else if (type == "Catalog")
+        schemaUrl = "https://schemas.stacspec.org/v1.0.0/catalog-spec/json-schema/catalog.json";
```
Comment (hobu) [inline io/StacReader.cpp:210]: @kylemann16 this URL should be a defaulted argument so the users can control which URL they want to try to apply schema validation too

File: io/StacReader.cpp [hunk 2]
```diff
@@ -0,0 +1,521 @@
...
+        NL::json args = m_readerArgs[driver].get<NL::json>();
+        for (auto& arg : args.items()) {
+            readerOptions.add(arg.key(), arg.value());
+        }
+    }
+
+    readerOptions.add("filename", dataUrl);
+    reader->setOptions(readerOptions);
+
+    if (m_readerList.size() > 0)
```
Comment (hobu) [inline io/StacReader.cpp:254]: If the m_readerList size is 0, is that an error?

File: io/StacReader.cpp [hunk 3]
```diff
@@ -0,0 +1,521 @@
...
+        reader->setInput(*m_readerList.back());
+
+    if (reader)
+    {
+        std::lock_guard<std::mutex> lock(m_mutex);
```
Comment (hobu) [inline io/StacReader.cpp:259]: Why are we locking here? Is this coming for the TIndex reader?

File: io/StacReader.cpp [hunk 4]
```diff
@@ -0,0 +1,521 @@
...
+        {
+            int value = properties[key].get<int>();
+            int desired = val.get<int>();
+            if (value != desired)
+                return false;
+            break;
+        }
+        case NL::detail::value_t::number_float:
+        {
+            int value = properties[key].get<int>();
```
Comment (hobu) [inline io/StacReader.cpp:312]: Is this bad copypasta? You're in the float case but fetching and asking for ints...

File: doc/stages/readers.stac.rst [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+Example
+--------------------------------------------------------------------------------
+
+.. code-block:: json
+[
+    {
+        "type": "readers.stac",
+        "filename": "https://s3-us-west-2.amazonaws.com/usgs-lidar-stac/ept/catalog.json",
+        "reader_args": [{"type": "readers.ept", "resolution": 100}],
+        "ids": ["MD_GoldenBeach_2012"],
```
[the remark this answers is not in the document]
Comment (hobu) [reply doc/stages/readers.stac.rst:20]: Also, point to std::regex docs as to the *type* of regex that is consumed.

File: io/StacReader.cpp (hunk 2 above)
Comment (kylemann16) [reply io/StacReader.cpp:254]: Not in this spot, but maybe elsewhere. Here we're just making sure that there are readers in the list before applying the previous back to the current reader's input.

File: io/StacReader.cpp (hunk 3 above)
Comment (kylemann16) [reply io/StacReader.cpp:259]: This is coming from [StageFactory.cpp](https://github.com/PDAL/PDAL/blob/master/pdal/StageFactory.cpp#L104-L114), which I was modeling this off of. I was thinking this reader should be multi threaded at that point, but didn't go through the rest of that. So I should probably remove it.

File: io/StacReader.cpp (hunk 4 above)
Comment (kylemann16) [reply io/StacReader.cpp:312]: Whoops, yes.

**Comment by hobu** [comment]
``readers.stac`` needs to be added to readers.rst list

File: doc/stages/readers.stac.rst [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+            "ids": ["MD_GoldenBeach_2012"],
+            "properties": { "pc:type": ["lidar", "sonar"], "pc:encoding": "ept" },
+            "asset_name": "ept.json",
+            "date_ranges": [
+                [
+                    "2022-11-11T0:00:0Z",
+                    "2022-11-30T0:00:0Z"
+                ]
+            ],
+            "schema_validate": true
```
Comment (hobu) [inline doc/stages/readers.stac.rst:32]: this got renamed

File: io/StacReader.cpp [hunk 7]
```diff
@@ -0,0 +1,587 @@
...
+        m_readerArgs[driver] = { };
+
+        for (auto& arg: readerPipeline.items())
+        {
+            if (arg.key() == "type")
+                continue;
+            m_readerArgs[driver][arg.key()] = arg.value();
+        }
+    }
+    std::cout << m_readerArgs.dump() << std::endl;
```
Comment (hobu) [inline io/StacReader.cpp:100]: clean up this lint

**Pull request closed without merging**